bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/AbstractHomekitAccessoryImpl.java

```diff
@@ -12,95 +12,160 @@
+        return accessory;
     }
 
-    protected String getItemName() {
-        return itemName;
+    public Collection<Service> getServices() {
+        return this.services;
     }
 
     protected HomekitAccessoryUpdater getUpdater() {
         return updater;
     }
 
-    protected GenericItem getItem() {
-        return (GenericItem) getItemRegistry().get(getItemName());
+    protected HomekitSettings getSettings() {
+        return settings;
+    }
+
+    protected void subscribe(HomekitCharacteristicType characteristicType,
+            HomekitCharacteristicChangeCallback callback) {
+        final Optional<HomekitTaggedItem> characteristic = getCharacteristic(characteristicType);
+        if (characteristic.isPresent()) {
+            getUpdater().subscribe((GenericItem) characteristic.get().getItem(), characteristicType.getTag(), callback);
+        } else {
+            LOGGER.error("Missing mandatory characteristic {}", characteristicType);
+        }
+    }
+
+    protected void unsubscribe(HomekitCharacteristicType characteristicType) {
+        final Optional<HomekitTaggedItem> characteristic = getCharacteristic(characteristicType);
+        if (characteristic.isPresent()) {
+            getUpdater().unsubscribe((GenericItem) characteristic.get().getItem(), characteristicType.getTag());
+        } else {
+            LOGGER.error("Missing mandatory characteristic {}", characteristicType);
+        }
+    }
+
+    protected @Nullable <T extends State> T getStateAs(HomekitCharacteristicType characteristic, Class type) {
+        final Optional<HomekitTaggedItem> taggedItem = getCharacteristic(characteristic);
+        if (taggedItem.isPresent() && taggedItem.get().getItem().getState() != null) {
+            return (T) taggedItem.get().getItem().getStateAs(type);
+        }
+        LOGGER.error("State for characteristic {} at accessory {} cannot be retrieved.", characteristic,
+                accessory.getId());
+
+        return null;
+    }
+
+    protected @Nullable <T extends Item> T getItem(HomekitCharacteristicType characteristic, Class<T> type) {
+        final Optional<HomekitTaggedItem> taggedItem = getCharacteristic(characteristic);
+        if (taggedItem.isPresent()) {
+            if (type.isInstance(taggedItem.get().getItem()))
+                return (T) getCharacteristic(characteristic).get().getItem();
+            else
+                LOGGER.error("Unsupported item type for characteristic {} at accessory {}. Expected {}, got {}",
+                        characteristic, accessory.getItem().getLabel(), type, taggedItem.get().getItem().getClass());
+        } else {
+            LOGGER.error("Mandatory characteristic {} not found at accessory {}. ", characteristic,
+                    accessory.getItem().getLabel());
+
+        }
+        return null;
+    }
+
+    protected @NonNull <T extends Object> T getAccessoryConfiguration(@NonNull String key, @NonNull T defaultValue) {
```

Comment:
I think it is already implied that T is an object
```suggestion
    protected @NonNull <T> T getAccessoryConfiguration(@NonNull String key, @NonNull T defaultValue) {
```